WordPress/src/main/java/org/wordpress/android/ui/posts/PostListMainViewModel.kt

```diff
@@ -419,6 +436,46 @@ class PostListMainViewModel @Inject constructor(
                 false
             }
 
+    fun handleRemotePreviewClosing() {
+        updatePreviewAndDialogState(PostListRemotePreviewState.NONE, PostInfoType.PostNoInfo)
+    }
+
+    private fun updatePreviewAndDialogState(newState: PostListRemotePreviewState, postInfo: PostInfoType) {
+        /* We need only transitions, so... */
+        if (_previewState.value == newState) return
+
+        AppLog.d(AppLog.T.POSTS, "****** POSTLIST PREVIEW SM - from [${_previewState.value}] to [$newState]")
+
+        /* update the state */
+        val prevState = _previewState.value
+        _previewState.postValue(newState)
+
+        /* take care of exit actions on state transition */
+        when (newState) {
+            PostListRemotePreviewState.PREVIEWING -> {
+                when (prevState) {
+                    PostListRemotePreviewState.UPLOADING_FOR_PREVIEW,
+                    PostListRemotePreviewState.REMOTE_AUTO_SAVING_FOR_PREVIEW -> {
+                        (postInfo as? PostInfoType.PostInfo)?.let { info ->
+                            info.post.let { post ->
+                                postActionHandler.handleRemotePreview(
+                                        localPostId = post.id,
+                                        remotePreviewType = if (prevState ==
+                                            PostListRemotePreviewState.UPLOADING_FOR_PREVIEW) {
+                                            RemotePreviewType.REMOTE_PREVIEW
+                                        } else {
+                                            RemotePreviewType.REMOTE_PREVIEW_WITH_REMOTE_AUTO_SAVE
+                                        }
+                                )
+                            }
+                        }
+                    }
+                }
+            }
+            else -> { /* nothing to do*/ }
```

Comment:
I think it is a more explicit and more "under control" approach and I like it. In general I guess could be a must when using 'when' (unwanted play on words 😊) as an expression in kotlin so you fully take advantage of kotlin error level checks (if not used as an expression AFAIU you get a warning I think). I would propose the following:
- in EditPostActivity.java I would list all the possible enum values to be explicit and add a default with throw exception.
- In PostListMainViewModel.kt maybe seems odd to list all possible cases and add an else (that AS reports as redundant) to throw an exception. So for now I (hopefully 😅) reached the goal managing the when as an expression body function (actually it could be an interesting pattern to decide to use when only as an, eventually dummy, expression in this or maybe better way 🤔). Also, rethinking to it, here for the nested when it actually seems to be more appropriate to use an if instead. Changes here d3958f7e6bd55072e65c999c3f0ebbf3e303a486 in `managePostLoadingStateTransitions` and `managePreviewStateTransitions`.

Reply:
>in EditPostActivity.java I would list all the possible enum values to be explicit and add a default with throw exception.

>In PostListMainViewModel.kt maybe seems odd to list all possible cases and add an else (that AS reports as redundant) to throw an exception.

> So for now I (hopefully 😅) reached the goal managing the when as an expression body function (actually it could be an interesting pattern to decide to use when only as an, eventually dummy, expression in this or maybe better way 🤔). 

Sorry, I should have been more explicit. I believe you can simply omit the `default/else` case if you list all the possible values. `Lint/ktlint` will abort the build if a `case` is missing (works on enums) -> which is even better than throwing an exception as you are informed about the issue before runtime. Wdyt?

Reply:
ok, I see 😊. Removed default branch in EditPostActivity.java. For PostListMainViewModel.kt since in any case I refactored to have a function, I staid with the expression body function but removed the comment since now is useless to underline that. Changes here d2dc6e2de721107aa3b5cfa9f1519cd1469b578f